### ISSUE TITLE
[Tracking] Sharded delta weight sync (delta_sharded): roadmap & known issues

### ISSUE THREAD

**Issue by ChangyiYang** [verl-project/verl#7060]
Tracking issue for the sharded delta weight sync (`delta_sharded` checkpoint-engine backend) merged in #6974, and the follow-up roadmap.

## Merged

- #6974 — `[ckpt,rollout] feat: sharded delta weight sync over NCCL for disaggregated rollout` (merged 2026-07-16, `903d90cc`)
  - Each trainer rank byte-diffs its **local shard** against a pinned-CPU snapshot; only changed `(position, value)` pairs are gathered to rank 0 and streamed to the rollout side. No rank ever stages a full-model mirror.
  - Shard layout is described declaratively via `ShardSpec` (`DeviceMesh` + `Placement`, `verl/workers/engine/spec.py`) — shared contract with the nccl-m2n work.
  - Measured steady-state sync vs `nccl` full broadcast (disaggregated, cross-node, offload=True): 7B **3.8s vs 9.1s (2.4×)**, 32B **12.5s vs 23.2s (1.9×)**, 72B **12.0s vs 36.9s (3.1×)**; 200-step GRPO equivalence run and independent re-measurement in the PR thread.

- #7144 — `ShardSpec` generalization: `BlockPlacement` block descriptors (any `Shard(k)` / multi-dim cuts) + explicit-place channel for hybrid geometries (merged 2026-07-27, `b9873ee9`)
- #7085 — veomni FSDP+EP shard export: fused expert stacks via a dim-0-separable converter (`to_hf_chunk` + handler-probed `hf_slots`), HSDP replica dedup (`ShardSpec.contributes`, covers newer VeOmni's `(ep_replicate, ep_fsdp, ep)` expert mesh), ep_size=1 degenerate geometry (merged 2026-07-29, `ce195c95`). Bit-level idempotence verify sweeps green at 30B (EP8 / EP1 / HSDP) and on VeOmni main.

## Next steps

- [x] **1. FSDP + EP: `Shard(1)` / block-descriptor placement** — DONE via #7144 + #7085. Original scope: — generalize `derive_placement` from the flat-offset `Shard(0)` fast path to a block descriptor `(local_shape, global_offset)` (both already provided by `compute_local_shape_and_global_offset`), with a vectorized local→global index translation and block placement in the dense first sync. Covers veomni-style FSDP+EP (`gate_up_proj` split on dim0 across EP ranks, local experts fully-sharded on dim1). Discussed in the #6974 review.</details>
- [x] **2. Megatron trainer support (bf16 rollout)** — DONE via #7181 (merged 2026-07-30, `48ca9eed`) (comm-stub probe over Megatron-Bridge mappings; TP+EP+ETP incl. hybrid-Mamba; bit-level verify sweeps + 235B 8.2–9.7× vs same-shape NCCL). PP is out of #7181's scope — see item 5. — mcore shard export through the same `ShardSpec` contract: 1-D TP `DeviceMesh` + `Shard(partition_dim)` + a pure-permutation `to_hf` converter closure; the engine's converter profile (boundary-preserving gather → NaN-sentinel rebuild → `to_hf` → HF-coordinate delta) already ships in #6974. (superseded working notes: https://github.com/ChangyiYang/verl/pull/2)
- [ ] **3. Quantized rollout (fp8)** — in progress (Megatron bf16+QAT trainer → sglang fp8 rollout, quant-domain delta): trainer-side sharded blockwise fp8 quantization (partial absmax grids + one `all_reduce(MAX)` per record, bitwise-identical to the rollout-side whole-tensor kernel), then byte-diff **in code space** — unchanged codes under an unchanged scale ship nothing, and measured per-128×128-block absmax flips are exactly zero across training windows at 7B/30B. Bit-level verify sweeps zero-mismatch with QAT enabled at 7B, 30B-A3B (EP8) and 235B-A22B (PP8×EP8); steady sync vs same-config fp8 NCCL: 7B **3.3–3.7s vs 5.0–5.6s**, 235B **62–79s vs 306–310s (4–5×)**. Both masked-copy loader guards from the known issue below (param-storage interval index; `post_load_weights` outside the patch) are implemented on this branch. PR to follow.
- [ ] **4. Composition with P2P / shard-level weight update (nccl-m2n)** — reuse the `ShardSpec` placement to route each shard's delta point-to-point to the rollout ranks that own it, instead of broadcasting the full delta stream to every worker.
- [ ] **5. Megatron pipeline parallelism (PP/VPP) for the delta export** — draft PR open: #7223 by @gxlvera (rides the bridge's global parameter directory; placeholder rows as zero-count lockstep entries; WORLD merge group with `contributes` dedup; row-keyed union slot tables also fixing an ep_size>1 expert misattribution. 235B TP4×PP8×EP4 verify sweep zero-mismatch, steady 2.1–2.4× vs same-shape NCCL).

## Known issue

- **Masked-copy guards in the SGLang delta loader.** The in-place sparse apply patches `Tensor.copy_` (scoped to the loader's `model.load_weights` calls) to skip NaN-masked positions. The slime implementation this design references — THUDM/slime#1806 (and its SGLang patch THUDM/slime#1993) — additionally (a) restricts the masked semantics to destinations whose storage belongs to `model.named_parameters()/named_buffers()` (a `data_ptr` interval index; non-param writes pass through untouched), and (b) runs `post_load_weights` / `process_weights_after_loading` in the unpatched environment so derived tensors (fp8 scales, MoE biases, MLA `w_kc`/`w_vc`) are recomputed normally. Neither guard is present in the merged loader. No impact observed in the shipped scope (bf16, standard dense models — the loaders there are pure slice/permute + `copy_`), but both guards should be ported before widening the scope (quant, models with post-load weight transforms).

- **(怀疑 / suspected)** The cupy staging pool used by the delta wire broadcast holds its blocks instead of returning them to CUDA: measured **4.15 GB** held after the full seed and ~0.6–0.9 GB after each steady sync (7B TP2 e2e; the engine logs `cupy staging pool after ... send: held X GB`). The engine now calls `free_all_blocks()` after every send as a mitigation; a proper review of the staging-buffer lifecycle for the fsdp and veomni paths is still owed.

**Comment by wuxibin89** [comment]
FSDP+EP expert placement could be one of `Shard(0)`, `Shard(1)`, `[Replicate(), Shard(0)]`, `[Replicate(), Shard(1)]`

**Comment by ChangyiYang** [comment]
## Perf matrix refresh (2026-07-20): dense 7B→72B, MoE 30B→235B, three engines

All numbers are steady-state `timing_s/update_weights` (seconds) on H100 nodes, GRPO/gsm8k smoke shapes, `delta_sharded` from the #7085 stack (block placement + sync-free masked apply) vs `nccl` from current main (incl. the #7005 staging fix — the older 7B off=T 9.1s / 32B off=T 23.2s nccl figures predate #7005 and are retired).

### Dense (FSDP2)

| model | offload | nccl | delta | speedup | notes |
|---|---|---|---|---|---|
| 7B | on | 5.5–6.0 | 3.9–4.9 | ~1.3× | gen_wait 2.4–2.7s = pageable H2D staging tax on both sides |
| 32B | on | 17.7–18.1 | 11.2–11.9 | 1.55× | delta seed 28s one-off |
| 32B | off | 14.2 | 6.2 | 2.3× | |
| 72B | off | 28.5–29.1 | 12.0–13.0 | 2.3× | 8 nodes, FSDP32, gen TP8; delta seed 41.6s; changed_ratio 2.1–2.6%, payload 8.6–10.9 GB; nccl sender: 23.5s of 24 is assembly/bucketing (83 buckets, 145 GB), wire_wait only 0.5–1.8s |

### MoE (three engines)

| model / engine | nccl | delta | speedup | notes |
|---|---|---|---|---|
| 30B-A3B, veomni ep8 | ~32 | 5.8–7.9 | **4.3×** | 3-step probe |
| 30B-A3B, veomni ep8, 50-step run | median 32.2 | median 7.1 | **4.5×** | equivalence run (below) |
| 30B-A3B, mcore TP4×EP2 | 24.6–25.2 | 17.1–19.7 | 1.4× | both sides dominated by per-expert conversion; the two batching knives tracked here should widen this |
| 235B-A22B, veomni ep8×fsdp8 | **255.9/246.5/266.1** | **14.9/11.8/11.4** | **≈21×** | 10 nodes (8 train + 2 rollout, gen TP16), offload on; delta changed_ratio 0.02–0.05%, payload 78–702 MB vs a ~470 GB bf16 full materialization per nccl sync |

### Training equivalence (30B-A3B, 50 steps each, wandb)

Same-seed GRPO pair, veomni ep8: score mean 0.682 (delta) vs 0.668 (nccl), rising 0.646→0.719 vs 0.639→0.697 — same-shape trajectories, per-step |Δ| ≈ 0.048 ≈ 3/64 of the sample granularity (independent-sampling noise, no drift). pg_loss/ppo_kl/grad_norm deltas are 1e-3–1e-1 absolute (noise around zero). Sync speedup holds for the whole run (median 7.1s vs 32.2s).

### Where the win comes from (unchanged conclusion, now measured up to 235B)

The delta path saves the *materialization done for the transfer*, not the wire time itself: sender-side full-tensor assembly (all-gathers) + bucket packing, per-bucket sync points and HTTP roundtrips (linear in bucket count), receiver-side bucket merges. With offload=on, both paths additionally pay a pageable H2D staging tax visible as gen_wait (2.5s @7B, ~13s @32B). The ratio therefore grows with model size and MoE sparsity: 1.3× (7B dense, offload) → 2.3× (32B/72B dense) → 4.3–4.5× (30B MoE) → ≈21× (235B MoE).

**Comment by 0z5a** [comment]
Update (2026-09-15): #7869 adds two opt-in GPU checks for `delta_sharded`: a focused CuPy staging-pool lifecycle test using real NCCL transfers, and a Qwen3-0.6B FSDP2 → SGLang GRPO smoke recipe covering training, weight sync, and rollout.

On two L20 GPUs, all 80 staging transfers matched exactly and the sender pool returned to zero after every release. The 10-step GRPO run completed with 10 zero-mismatch weight-verification sweeps. No persistent staging retention was reproduced in these bounded runs. The PR changes tests only; reproduction commands, environment details, and coverage limits are in #7869.

The earlier gather-probe exit-status fix was submitted separately as #7776 (now closed); its validation details remain in that PR.

**Comment by 0z5a** [comment]
Follow-up validation of the staging-pool item on unmodified main `61134f92099913b6f8fa5920ac5839a4a314212d`: a real two-L20 FSDP2 → SGLang GRPO smoke run completed successfully.

This is additional bounded test evidence, not a proposed production fix or a claim that the full tracking issue is resolved.

### Setup and results

- Qwen/Qwen3-0.6B, revision `c1899de289a04d12100db370d81485cdf75e47ca`; one FSDP2 training rank and one independent SGLang rollout rank (TP1), one L20 per rank.
- Real generation → deterministic synthetic reward → GRPO backward/optimizer step → backend delta export → ZMQ/NCCL transfer → production SGLang weight loader → subsequent generation.
- Independent 5-step and 10-step executions both exited 0. The 10-step execution recorded **10 `DELTA-VERIFY` sweeps, each with 311 manifest parameter items, `mismatch_elems=0 offenders=[]`**. Successful verification logs were hidden by the default server error log level in the first execution; the second explicitly enabled warning logs.
- All logged losses/gradient norms were finite. In steps 2–10, gradient norms were 5.45–10.74 and changed weight ratios were 4.04%–7.64%; this was not only a zero-update test.
- Delta payload: 173.79–328.68 MiB, excluding the additional dense verification sweep. Weight-sync wall time **including verification**: 0.870–0.927 s (median 0.890 s). These are diagnostic timings, not an A/B speedup claim.

### Staging observations

The initial seed pool held 1.40 GiB before release; device free memory increased from 37.12 to 38.52 GiB after `free_all_blocks()`. Later verification sends held about 1.57–1.58 GiB (dense verification plus cached delta blocks); release returned device free memory to approximately 30.08–30.09 GiB, and the following steady-release log reported 0.00 GiB held. Trainer Torch allocated/reserved peaks were 11.121/13.549 GiB. This bounded run did not reproduce persistent pool retention. These production log values are rounded, not byte-exact allocator measurements.

A separate staging/transport probe ran 80 sends across two fresh actor pairs, alternating seed/zero/sparse/near-full payloads. Received values/positions matched exactly and all 80 sender post-release CuPy `used_bytes()` / `total_bytes()` samples were zero. That probe is separate evidence, not the GRPO E2E itself.

### Reproduction

Environment: Torch 2.13.0+cu130, SGLang 0.5.19, Transformers 5.10.2, TensorDict 0.10.0. This differs from the repository's pinned Torch 2.11/SGLang 0.5.12 environment. Use the exact source revision above, with the following files in `/experiment` and the model cached at the path in the scripts. Both runs were bounded by a 1200-second host timeout.

<details>
<summary>prepare_e2e.py</summary>

```python
from pathlib import Path
import pandas as pd
from transformers import AutoTokenizer
model="/root/.cache/huggingface/hub/models--Qwen--Qwen3-0.6B/snapshots/c1899de289a04d12100db370d81485cdf75e47ca"
AutoTokenizer.from_pretrained(model,local_files_only=True)
rows=[dict(data_source="delta_smoke",prompt=[dict(role="user",content=f"Calculate {i%8}+{(i+1)%8}. Reply with the result.")],ability="math",reward_model=dict(style="rule",ground_truth=str(i%8+(i+1)%8)),extra_info=dict(index=i)) for i in range(32)]
pd.DataFrame(rows).to_parquet("/experiment/e2e_train.parquet")
pd.DataFrame(rows[:4]).to_parquet("/experiment/e2e_val.parquet")
print("INPUTS_OK",model)
```
</details>

<details>
<summary>e2e_reward.py</summary>

```python
def compute_score(data_source, solution_str, ground_truth, extra_info=None, **kwargs):
    # A deterministic bounded synthetic reward to exercise real policy updates.
    # This is a plumbing validation, not a math-quality evaluation.
    return float(sum(solution_str.encode("utf-8")) % 17) / 16.0
```
</details>

<details>
<summary>run_e2e.sh</summary>

```bash
#!/bin/bash
set -euo pipefail
cd /experiment
export PYTHONPATH=/experiment
export OMP_NUM_THREADS=4
export TOKENIZERS_PARALLELISM=false
export HYDRA_FULL_ERROR=1
export RAY_DEDUP_LOGS=0
export HF_HUB_OFFLINE=1
python3 -m verl.experimental.one_step_off_policy.main_ppo \
 actor_rollout_ref.model.path=/root/.cache/huggingface/hub/models--Qwen--Qwen3-0.6B/snapshots/c1899de289a04d12100db370d81485cdf75e47ca \
 data.train_files=/experiment/e2e_train.parquet data.val_files=/experiment/e2e_val.parquet \
 data.train_batch_size=4 data.max_prompt_length=128 data.max_response_length=32 \
 data.filter_overlong_prompts=True data.truncation=error \
 actor_rollout_ref.actor.strategy=fsdp2 actor_rollout_ref.hybrid_engine=False \
 actor_rollout_ref.actor.ppo_mini_batch_size=4 actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=1 \
 actor_rollout_ref.actor.optim.lr=1e-5 actor_rollout_ref.actor.use_kl_loss=False \
 actor_rollout_ref.actor.fsdp_config.use_torch_compile=False \
 actor_rollout_ref.model.use_remove_padding=False actor_rollout_ref.model.enable_gradient_checkpointing=True \
 +actor_rollout_ref.model.override_config.attn_implementation=sdpa \
 actor_rollout_ref.rollout.name=sglang actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
 actor_rollout_ref.rollout.n=2 actor_rollout_ref.rollout.gpu_memory_utilization=0.35 \
 actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=1 \
 actor_rollout_ref.rollout.max_model_len=256 actor_rollout_ref.rollout.agent.num_workers=2 \
 +actor_rollout_ref.rollout.engine_kwargs.sglang.attention_backend=triton \
 +actor_rollout_ref.rollout.engine_kwargs.sglang.disable_cuda_graph=True \
 actor_rollout_ref.rollout.checkpoint_engine.backend=delta_sharded \
 +actor_rollout_ref.rollout.checkpoint_engine.engine_kwargs.delta_sharded.encoding=indices \
 +actor_rollout_ref.rollout.checkpoint_engine.engine_kwargs.delta_sharded.verify_every=1 \
 algorithm.adv_estimator=grpo algorithm.use_kl_in_reward=False \
 reward.num_workers=2 reward.custom_reward_function.path=/experiment/e2e_reward.py \
 reward.custom_reward_function.name=compute_score \
 trainer.nnodes=1 trainer.n_gpus_per_node=1 rollout.nnodes=1 rollout.n_gpus_per_node=1 \
 trainer.total_epochs=1 trainer.total_training_steps=5 trainer.val_before_train=False \
 trainer.save_freq=-1 trainer.test_freq=-1 trainer.logger='[console]' \
 trainer.project_name=delta-l20-validation trainer.experiment_name=fsdp2-sglang \
 ray_kwargs.ray_init.num_cpus=32 "$@"
```
</details>

```bash
cd /experiment
python3 prepare_e2e.py
bash run_e2e.sh
bash run_e2e.sh trainer.total_training_steps=10 trainer.total_epochs=2 trainer.experiment_name=fsdp2-sglang-verify +actor_rollout_ref.rollout.engine_kwargs.sglang.log_level=warning
```

### Limits

One training rank does not cover multi-rank FSDP shard assembly, VeOmni/EP, multi-node runs, interrupted collectives or long-duration allocator drift. The native one-step-off-policy schedule verifies each performed sync; no extra sync was added after the final optimizer step. There was no independent full-NCCL output-parity baseline. Responses reached the deliberately short 32-token cap, so this is plumbing/correctness evidence, not model-quality or convergence evidence.

With warning logging enabled, SGLang emitted a nonfatal startup `/freeze_gc` connection refusal to the default port 30000, then served the configured endpoint and completed all checks. The native tied-embedding loader also warned that `lm_head.weight` was absent from its unique parameter dictionary. These are not being represented as delta mismatches or silently omitted environment observations.

No new production change is proposed based on these results, and the remaining FSDP/VeOmni lifecycle investigation is still open.
